README.md
# Jsonmatic

Transform a CSV into a JSON.

You have:

| key | road | coord.lat | coord.lng | elem |
|-----|------|-----------|-----------|------|
| 1   | AP-7 | 42.02     | 2.82      | 🦄    |
| 2   | C-32 | 41.35     | 2.09      | 🦧    |
| 3   | B-20 | 41.44     | 2.18      | 🐰    |

You get:

```javascript
{
    "1": {
        "road": "AP-7",
        "coord": {
            "lat": 42.02,
            "lng": 2.82
        },
        "elem": "🦄'"
    },
    "2": {
        "road": "C-32",
        "coord": {
            "lat": 41.35,
            "lng": 2.09
        },
        "elem": "🦧"
    },
    "3": {
        "road": "B-20",
        "coord": {
            "lat": 41.44,
            "lng": 2.18
        },
        "elem": "🐰"
    }
}
```

## Author

[Erik Martín Jordán](https://erikmartinjordan.com)

## License

This project is open source and available under the MIT License.


src/changelog.md
## 04.06.21 – New copy/download options and bug fixing

* Fixed splitting on reading a CSV
* Updated textarea font to `monospace` as suggested on HN
* Removed `fingerprintjs` and analytics component
* Added CSV copy/download options
* Replaced dot separator with `__separator__` on TransformToCSV component
* Autofocus on table cell

## 03.06.21 – Upload CSV

* Option to upload CSV instead of just copy/paste
* Fixed textarea on mobile devices 

## 01.06.21 – Randomize JSON

* Create a dummy JSON

## 24.05.21 – Keyboard shortcut to upload files

* Added keyboard control to upload files
* Added dummy CSV animation when transforming JSON
* Deleted '+' sign on keyboard shortcuts hint because it can create confusion
* Added animation to keyboard hint on 'Upload' component

## 21.05.21 — Resize table

![Resize table](https://github.com/erikmartinjordan/Screenshots/blob/master/May-21-2021%2011-31-56.gif?raw=true)

* Resizable table
* Autoresize on double click
* Disabled "Generate CSV" button while editing JSON

## 20.05.21 — Transform JSON into CSV

* Transform JSON into CSV

## 12.05.21 — JSON edition

![JSON edition](https://github.com/erikmartinjordan/Screenshots/blob/master/May-12-2021%2013-05-23.gif?raw=true)

* Edit JSON textarea
* Buttons are disabled while JSON edition
* Fixed bug: listener of keyboard active when some table cell is selected

## 05.05.21 — Multiple JSON merge

![Merge JSON files](https://github.com/erikmartinjordan/Screenshots/blob/master/May-05-2021%2012-47-05.gif?raw=true)

* Merge multiple JSON files
* Select merge order (priority)
* Download merged file

## 02.05.21 — Multiple JSON deletion

* Delete several properties of multiple JSON files separating them with commas: `prop1.subprop1`, `prop2.subprop2`, etc.
* Delete a single property using rules (*equal to*, *greater than*, and *lesser than*)
* Displaying number of deletions

## 30.04.21 — Rules

* Replace fields if a property is greater/lesser than a certain value
* Added tilt effect on hover buttons 💅
* Added file name as title into JSON box 
* Displaying number of replaces

## 29.04.21 — Multiple JSON replacement

![Multiple JSON replacement](https://github.com/erikmartinjordan/Screenshots/blob/master/May-05-2021%2012-46-00.gif?raw=true)

* Added a component to replace a property in multiple JSON files
* Replace/download

## 27.04.21 — Changelog

* Added changelog

## 17.03.21 — Launch 🚀

Transform a CSV into a JSON automatically:

* Copy a CSV and paste it
* Add/delete columns & rows
* Undo/redo options
* Keyboard shortcuts
* Download the JSON as file or copy it
* Beautify/uglify the JSON
* 2-space/4-space indentation

src/App.js
import React, { useState }  from 'react';
import useUndoableState     from './useUndoableState';
import Header               from './Header';
import Json                 from './Json';
import Csv                  from './Csv';
import TransformToJSON      from './TransformToJSON';
import TransformToCSV       from './TransformToCSV';
import Menu                 from './Menu';

const App = () => {
    
    const [csv, setCsv, undo, redo] = useUndoableState([
        
        ['key', 'road', 'coord.lat',  'coord.lng', 'elem'],
        ['1',   'AP-7', 42.02,        2.82,        '🦄'],
        ['2',   'C-32', 41.35,        2.09,        '🦧'],
        ['3',   'B-20', 41.44,        2.18,        '🐰'],
        
    ]);
    
    const [json, setJson] = useState(
        
        {
            "1": {
                "road": "AP-7",
                "coord": {
                    "lat": 42.02,
                    "lng": 2.82
                },
                "elem": "🦄"
            },
            "2": {
                "road": "C-32",
                "coord": {
                    "lat": 41.35,
                    "lng": 2.09
                },
                "elem": "🦧"
            }, 
            "3": {
                "road": "B-20",
                "coord": {
                    "lat": 41.44,
                    "lng": 2.18
                },
                "elem": "🐰"
            }
        }
        
    );
    
    const [select, setSelect] = useState(['', '', '', '']);
    const [edit, setEdit]     = useState(false);
    
    return (
        <div className = 'App'>
            <Header/>
            <div className = 'Content'>
                <Csv 
                    csv       = {csv}
                    setCsv    = {setCsv}
                    select    = {select}
                    setSelect = {setSelect}
                />
                <div className = 'Buttons'>
                    <TransformToJSON
                        csv     = {csv}
                        json    = {json} 
                        setJson = {setJson}    
                    />
                    <TransformToCSV
                        json    = {json} 
                        csv     = {csv}
                        edit    = {edit}
                        setCsv  = {setCsv}    
                    />
                </div>
                <Json
                    edit    = {edit} 
                    json    = {json}
                    setEdit = {setEdit}
                    setJson = {setJson}
                />
            </div>
            <Menu
                csv         = {csv}
                setCsv      = {setCsv}
                select      = {select}
                setSelect   = {setSelect}   
                undo        = {undo}
                redo        = {redo}
            />
        </div>
    );
    
}

export default App;